src/freezethaw.rs

```diff
@@ -0,0 +1,58 @@
+use linux_raw_sys::ioctl::FIFREEZE;
+use linux_raw_sys::ioctl::FITHAW;
+
+use rustix::fd::AsFd;
+use rustix::{io, ioctl};
+
+fn ioctl_fifreeze<Fd: AsFd>(fd: Fd) -> io::Result<()> {
+    // SAFETY: `FIFREEZE` is a no-argument opcode.
+    unsafe {
+        let ctl = ioctl::NoArg::<{ FIFREEZE }>::new();
+        ioctl::ioctl(fd, ctl)
+    }
+}
+
+fn ioctl_fithaw<Fd: AsFd>(fd: Fd) -> io::Result<()> {
+    // SAFETY: `FITHAW` is a no-argument opcode.
+    unsafe {
+        let ctl = ioctl::NoArg::<{ FITHAW }>::new();
+        ioctl::ioctl(fd, ctl)
+    }
+}
+
+/// syncfs() doesn't flush the journal on XFS,
+/// and since GRUB2 can't read the XFS journal, the system
```

Comment:
Do we really know it's only xfs? It seems safer to do this unconditionally.

Reply:
Done, ignoring `EOPNOTSUPP` and `EPERM`